qiskit_experiments/curve_analysis/scatter_table.py

```diff
@@ -9,185 +9,498 @@
+        mini_y = table.yvals(data_uid="model1", category="raw", analysis="Analysis_A")
+
+    When an analysis only has a single model and the table is created from a single
+    analysis instance, the data_uid and analysis are trivial, and you only need to
+    specify the category to get subset data of interest.
 
     """
 
-    # TODO Add this to toctree. In current mechanism all pandas DataFrame members are rendered
-    #  and it fails in the Sphinx build process. We may need a custom directive to
-    #  exclude class members from an external package.
+    DEFAULT_COLUMNS = [
+        "xval",
+        "yval",
+        "yerr",
+        "name",
+        "data_uid",
+        "category",
+        "shots",
+        "analysis",
+    ]
 
-    @classmethod
-    def _default_columns(cls) -> List[str]:
-        return [
-            "xval",
-            "yval",
-            "yerr",
-            "name",
-            "class_id",
-            "category",
-            "shots",
-        ]
+    DEFAULT_DTYPES = [
+        "Float64",
+        "Float64",
+        "Float64",
+        "string",
+        "Int64",
+        "string",
+        "Int64",
+        "string",
+    ]
 
-    @deprecate_func(
-        since="0.6",
-        additional_msg="Curve data uses dataframe representation. Use dataframe filtering method.",
-        pending=True,
-        package_name="qiskit-experiments",
-    )
-    def get_subset_of(self, index: Union[str, int]) -> "ScatterTable":
-        """Filter data by series name or index.
+    def __init__(self):
+        self._lazy_add_rows = []
+        self._dump = pd.DataFrame(columns=self.DEFAULT_COLUMNS)
+
+    @property
+    def _data(self) -> pd.DataFrame:
+        if self._lazy_add_rows:
+            # Add data when table element is called.
+            # Adding rows in loop is extremely slow in pandas.
+            tmp_df = pd.DataFrame(self._lazy_add_rows, columns=self.DEFAULT_COLUMNS)
+            tmp_df = self._format_table(tmp_df)
+            if len(self._dump) == 0:
+                self._dump = tmp_df
+            else:
+                self._dump = pd.concat([self._dump, tmp_df], ignore_index=True)
+            self._lazy_add_rows.clear()
+        return self._dump
+
+    @classmethod
+    def from_dataframe(cls, data: pd.DataFrame) -> "ScatterTable":
+        """Create new dataset with existing dataframe.
 
         Args:
-            index: Series index of name.
+            data: Data dataframe object.
 
         Returns:
-            A subset of data corresponding to a particular series.
+            A new ScatterTable instance.
         """
-        if isinstance(index, int):
-            index = self.labels[index]
-        return self[self.name == index]
+        if list(data.columns) != cls.DEFAULT_COLUMNS:
+            raise ValueError("Input dataframe columns don't match with the ScatterTable spec.")
+        instance = object.__new__(ScatterTable)
```

Comment:
This is to avoid creation of unnecessary dataframe in the constructor. Alternatively we can do `def __init__(self, data: pd.DataFrame):` instead of the factory method.

Reply:
Okay, hmm, maybe an `__init__` argument would be cleaner though I don't have a strong preference. With the current way `from_dataframe` needs to be kept in sync with the behavior of `__init__`. `from_dataframe` is used in several tests, so we should catch if something is added to `__init__` but not here.